apps/api_web/lib/api_web/controllers/route_pattern_controller.ex

```diff
@@ -92,25 +97,130 @@ defmodule ApiWeb.RoutePatternController do
   defp reject_invalid_canonical_filter(filters), do: filters
 
   @spec format_filters(%{optional(String.t()) => String.t()}) :: RoutePattern.filters()
-  defp format_filters(filters) do
-    Map.new(filters, fn {key, value} ->
-      case {key, value} do
-        {"id", ids} ->
-          {:ids, Params.split_on_comma(ids)}
+  defp format_filters(filters, acc \\ %{})
 
-        {"route", route_ids} ->
-          {:route_ids, Params.split_on_comma(route_ids)}
+  defp format_filters(%{"id" => ids} = filters, acc) do
+    new_acc = Map.put(acc, :ids, Params.split_on_comma(ids))
 
-        {"direction_id", direction_id} ->
-          {:direction_id, Params.direction_id(%{"direction_id" => direction_id})}
+    filters
+    |> Map.delete("id")
+    |> format_filters(new_acc)
+  end
+
+  defp format_filters(%{"direction_id" => direction_id} = filters, acc) do
+    new_acc = Map.put(acc, :direction_id, Params.direction_id(%{"direction_id" => direction_id}))
+
+    filters
+    |> Map.delete("direction_id")
+    |> format_filters(new_acc)
+  end
+
+  defp format_filters(%{"route" => route_ids} = filters, acc) do
+    new_acc = Map.put(acc, :route_ids, Params.split_on_comma(route_ids))
+
+    filters
+    |> Map.delete("route")
+    |> format_filters(new_acc)
+  end
+
+  defp format_filters(%{"date" => date} = filters, acc) do
+    case process_date(date, acc) do
+      :error ->
+        filters
+        |> Map.delete("date")
+        |> format_filters(acc)
+
+      new_acc ->
+        filters
+        |> Map.delete("date")
+        |> format_filters(new_acc)
+    end
+  end
+
+  defp format_filters(%{"stop" => stop_ids} = filters, acc) do
+    new_acc = Map.put(acc, :stop_ids, Params.split_on_comma(stop_ids))
+
+    filters
+    |> Map.delete("stop")
+    |> format_filters(new_acc)
+  end
+
+  defp format_filters(%{"canonical" => canonical} = filters, acc) do
+    new_acc = Map.put(acc, :canonical, Params.canonical(canonical))
+
+    filters
+    |> Map.delete("canonical")
+    |> format_filters(new_acc)
+  end
+
+  defp format_filters(_, acc) do
+    acc
+  end
 
-        {"stop", stop_ids} ->
-          {:stop_ids, Params.split_on_comma(stop_ids)}
+  defp consolidate_route_ids(%{route_ids: [_ | _] = route_ids}, [_ | _] = date_specific_route_ids) do
+    route_ids
+    |> MapSet.new()
+    |> MapSet.intersection(MapSet.new(date_specific_route_ids))
+    |> MapSet.to_list()
+  end
+
+  defp consolidate_route_ids(_, date_specific_route_ids), do: date_specific_route_ids
+
+  defp gather_trip_based_route_pattern_ids(
+         %{ids: ids, date: _date, route_ids: [_ | _] = route_ids} = acc
+       ) do
+    acc
+    |> Map.take([:date, :direction_id, :route_patterns])
+    |> Map.put(:route_patterns, ids)
```

Comment:
```suggestion
    |> Map.take([:date, :direction_id])
    |> Map.put(:route_patterns, ids)
```
suggestion: it looks like you're taking `route_patterns` and them immediately overwriting it.